coderd/notifications/notifier.go

```diff
@@ -0,0 +1,233 @@
+		Count:           int32(n.cfg.LeaseCount),
+		MaxAttemptCount: int32(n.cfg.MaxSendAttempts),
+		NotifierID:      n.id,
+		LeaseSeconds:    int32(n.cfg.LeasePeriod.Value().Seconds()),
+	})
+	if err != nil {
+		return nil, xerrors.Errorf("acquire messages: %w", err)
+	}
+
+	return msgs, nil
+}
+
+// prepare has two roles:
+// 1. render the title & body templates
+// 2. build a dispatcher from the given message, payload, and these templates - to be used for delivering the notification
+func (n *notifier) prepare(ctx context.Context, msg database.AcquireNotificationMessagesRow) (dispatch.DeliveryFunc, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+
+	// NOTE: when we change the format of the MessagePayload, we have to bump its version and handle unmarshalling
+	// differently here based on that version.
+	var payload types.MessagePayload
+	err := json.Unmarshal(msg.Payload, &payload)
+	if err != nil {
+		return nil, xerrors.Errorf("unmarshal payload: %w", err)
+	}
+
+	handler, ok := n.handlers[msg.Method]
+	if !ok {
+		return nil, xerrors.Errorf("failed to resolve handler %q", msg.Method)
+	}
+
+	var title, body string
+	if title, err = render.GoTemplate(msg.TitleTemplate, payload, nil); err != nil {
+		return nil, xerrors.Errorf("render title: %w", err)
+	}
+	if body, err = render.GoTemplate(msg.BodyTemplate, payload, nil); err != nil {
+		return nil, xerrors.Errorf("render body: %w", err)
+	}
+
+	return handler.Dispatcher(payload, title, body)
+}
+
+// deliver sends a given notification message via its defined method.
+// This method *only* returns an error when a context error occurs; any other error is interpreted as a failure to
+// deliver the notification and as such the message will be marked as failed (to later be optionally retried).
+func (n *notifier) deliver(ctx context.Context, msg database.AcquireNotificationMessagesRow, deliver dispatch.DeliveryFunc, success, failure chan<- dispatchResult) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, n.cfg.DispatchTimeout.Value())
+	defer cancel()
+	logger := n.log.With(slog.F("msg_id", msg.ID), slog.F("method", msg.Method))
+
+	retryable, err := deliver(ctx, msg.ID)
+	if err != nil {
+		// Don't try to accumulate message responses if the context has been canceled.
+		//
+		// This message's lease will expire in the store and will be requeued.
+		// It's possible this will lead to a message being delivered more than once, and that is why Stop() is preferable
+		// instead of canceling the context.
+		//
+		// In the case of backpressure (i.e. the success/failure channels are full because the database is slow),
+		// we can't append any more updates to the channels otherwise this, too, will block.
+		if xerrors.Is(err, context.Canceled) {
+			return err
+		}
+
+		logger.Warn(ctx, "message dispatch failed", slog.Error(err))
+		failure <- newFailedDispatch(n.id, msg.ID, err, retryable)
```

Comment:
We need to be careful about writes to `failure` and `success` blocking in the case of an ungraceful shutdown.  If either are full we will leak delivery goroutines, since the manager stops reading from these channels when the context is canceled.

I realize that you're checking for context canceled errors above, but there is no guarantee that `deliver()` actually bothers to check the context and return an error.  Even if it does check, there is still a race condition where `deliver()` succeeds (or fails without `context.Canceled`), then the manager exits and stops reading from the channel, then we write to the channel and block forever.

The simplest thing to do is put the writes in a select with a read from `context.Done()`